proposals/scopes.md

```diff
@@ -196,106 +196,179 @@ interface OriginalPosition {
 
 ### Encoding
 
-We introduce two new fields "originalScopes" and "generatedRanges" respectively:
+We introduce a new field "scopes" to the source map JSON: "scopes" is a string. It contains a list of comma-separated items. Each item is prefixed with a unique "tag". The items themselves build a tree structure that describe "original scope" and "generated range" trees.
 
-  * "originalScopes" is an array of original scope tree descriptors (a string). Each element in the array describes the scope tree of the corresponding "sources" entry.
-  * "generatedRanges" is a single generated range tree descriptor (a string) of the generated file.
+The format of "scopes" is presented in an EBNF-like grammar, with:
 
-Like the "mappings" field, the data in a "generated range tree descriptor" is grouped by line and lines are separated by `;`. Within a line, items are separated by `,`. A "original scope tree descriptor" is NOT grouped by line, but items are separated by `,`.
+  * Three terminals: Signed, unsigned VLQ and comma '`,`'. VLQ terminals are labelled and we denote them with UPPERCASE.
+    We prefix the terminal with `u` or `s` to signify an unsigned or signed VLQ respectively. E.g. the terminal `uLINE` signifies
+    an unsigned VLQ labelled `LINE`.
 
-There are two different kinds of items that will appear in a "original scope descriptor": "Start Original Scope" and "End Original Scope".
-There are two different kinds of items that will appear in the "generated range descriptor": "Start Generated Range" and "End Generated Range".
+  * Non-terminals are denoted with `snake_case`.
 
-The kind of an item can be determined by looking at how many VLQ-encoded numbers it contains: "End" items contain one number, "Start" items contain two or more numbers.
+  * `symbol?` means zero or one `symbol`.
+  * `symbol+` means one or more `symbol`.
 
-Note: Each DATA represents one VLQ number.
+The start symbol is `scopes`:
 
-#### Start Original Scope
+```
+scopes := original_scope_list generated_range_list
+
+original_scope_list :=
+    original_scope_tree
+  | ε
+  | original_scope_list ',' original_scope_tree
+  | original_scope_list ',' ε
+
+generated_range_list :=
+    generated_range_tree
+  | generated_range_list ',' generated_range_tree
+```
+
+Only one top-level `original_scope_tree` per `sources` file is allowed. The n-th top-level `original_scope_tree` describes the scope tree for `sources[n]`. To signify that a certain `sources[m]` authored file doesn't have scopes information available, an empty item must be used.
+Multiple top-level `generated_range_tree`s are allowed, this is especially useful when multiple bundles are straight-up concatenated.
+
+#### Original Scope Trees
+
+```
+original_scope_tree :=
+  original_scope_start
+  original_scope_variables?
+  original_scope_source_index?
+  original_scope_tree?
```

Comment:
There can be multiple `original_scope_tree` children.

Reply:
Nice catch, I'll update it to be a `original_scope_tree*`.

Reply:
I aligned it with the latest proposal spec text, so we use `start ',' item_list? end`. And the `item_list` can be an arbitrary number of variables or trees.

In the final version of the grammar we might want to restrict `original_scope_variables` to one occurrence at most per start/end pair, but that would make it harder to read for now.